### DISCUSSION TITLE
SEVI速度太慢，请问怎么优化

### DISCUSSION THREAD

**Discussion by meishuyao** [physics-data/faq#120] [Q&A] [answered]
sevi的运行速度太慢，请问优化有什么比较好的办法吗

**Comment by heroxbd** [comment, marked as answer]
提升程序的运行速度，需要先测量耗时最长的部分，以便 有针对性地进行优化。这一过程叫做 `profile`。

Python 的 `profile` 工具是 `cProfile`，在 Python 的文档中有述

  https://docs.python.org/3/library/profile.html

cProfile 会生成一个统计文件，有很多查看器可以读取该文件。常见的查看器有 gprof2dot[1] ，可以画出程序的时间占用图。

1. https://github.com/jrfonseca/gprof2dot

**Comment by meishuyao** [comment]
运行了统计程序之后发现大部分时间花在了数据压缩上，这是不是意味着没有什么改进空间了
@heroxbd

**Reply by heroxbd** [reply]
有图吗？可以调节压缩的级别。

**Reply by meishuyao** [reply]
我已经把压缩的级别逐级试过了，找到了时间和空间比较平衡的级别
接下来可能的优化是不是只有并行化了

**Comment by heroxbd** [comment]
meishuyao ***@***.***> writes:

> 我已经把压缩的级别逐级试过了，找到了时间和空间比较平衡的级别
> 接下来可能的优化是不是只有并行化了

直觉上讲是矢量优化，从 profile 结果看它真的不是主要因素吗？

另外，你的文件系统是虚拟机里的吗，会不会瓶颈在文件系统的写入性能？

**Reply by meishuyao** [reply]
我没找到这个discussion怎么插入图片。我运行的是基础版本的模拟，没有level123。有85%的时间花在h5py.create_dset上，我改变压缩等级后这个比例也会变。另外我用的电脑是linux单系统，机械硬盘没有固态。

**Reply by meishuyao** [reply]
现在还在写level123的优化

**Comment by littlewu2508** [comment]
有关 HDF5 写入大文件的 hint：

向 hdf5 中写入大量数据时，最合适的方案是将数据分为合适大小的若干 block，然后依次将每个 block 的数据从内存写入硬盘。

比如，对于 50000 张图，可以每次计算 200 张图存入内存，然后 200 张一起存入硬盘。如果一张一张存，会对硬盘开销大，尤其会拖慢机械硬盘；此外，hdf5 的压缩对更大块的数据效率也更高，所以 chunk size 大一些压缩效果会略好。

具体实现，以 h5py 为例，参考 [create_dataset](https://docs.h5py.org/en/stable/high/group.html?highlight=craete_dataset#h5py.Group.create_dataset)，可以在创建 dataset 时手动指定表格的 size（例如 50000行，还可以指定其它参数，例如 chunks），然后把这个 dataset 当做一个存在于在硬盘上数组，按照 https://docs.h5py.org/en/stable/high/dataset.html#reading-writing-data 的方法每次存 200x2048x2048 的数据，h5py 会一次性将 200 张图存入表格。这样做对于硬盘的 IO 开销小，且始终只有 200 张图占用内存。
